### PR TITLE
feat(dashboards): Added query selector to dashboard open in discover button

### PR DESCRIPTION
**Pull request by edwardgou-sentry**
Updated the Open in Discover button in Dashboard Widgets to display a query selector Modal. This allows the user to select which query from the widget they would like to view in discover.
image

### CODE REVIEW

File: static/app/views/dashboardsV2/widgetCard.tsx [hunk 1]
```diff
@@ -140,7 +130,7 @@ class WidgetCard extends React.Component<Props> {
                 eventName: 'Dashboards2: Table Widget - Open in Discover',
                 organization_id: parseInt(this.props.organization.id, 10),
               });
-              browserHistory.push(eventView.getResultsViewUrlTarget(organization.slug));
+              openDashboardWidgetQuerySelectorModal({organization, widget});
```
Comment (shruthilayaj) [inline static/app/views/dashboardsV2/widgetCard.tsx:133]: This opens the dashboard selector modal for table/big number etc widgets which can only have one query and for chart widgets with a single query. We should have these go straight to discover right? Not sure what the design is on that, but it doesn't match the "Multiple queries were used to create this widget visualization. Which query would you like to view in Discover?" description in the modal

Comment (shruthilayaj) [reply static/app/views/dashboardsV2/widgetCard.tsx:133]: Also, this should be feature gated by the `connect-discover-and-dashboards` flag for the table widget right?

File: tests/js/spec/components/modals/dashboardWidgetQuerySelectorModal.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    await tick();
+    expect(wrapper.find('StyledInput').length).toEqual(1);
+    expect(wrapper.find('StyledInput').props().value).toEqual(
+      'title:/organizations/:orgId/performance/summary/'
+    );
+    expect(wrapper.find('OpenInDiscoverButton').length).toEqual(1);
+    wrapper.unmount();
+  });
+
+  it('renders a multiple query selections when the widget only has multiple querys', async function () {
```
Comment (shruthilayaj) [inline tests/js/spec/components/modals/dashboardWidgetQuerySelectorModal.spec.tsx:109]: nit:
```suggestion
  it('renders a multiple query selections when the widget only has multiple queries', async function () {
```

File: static/app/components/modals/dashboardWidgetQuerySelectorModal.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  organization: Organization;
+  widget: Widget;
+};
+
+type Props = ModalRenderProps &
+  DashboardWidgetQuerySelectorModalOptions & {
+    api: Client;
+    organization: Organization;
+    selection: GlobalSelection;
+    tags: TagCollection;
```
[the remark this answers is not in the document]
Comment (edwardgou-sentry) [reply static/app/components/modals/dashboardWidgetQuerySelectorModal.tsx]: 😅  I thought I removed these. Updated!

File: static/app/components/modals/dashboardWidgetQuerySelectorModal.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  color: ${p => p.theme.gray300};
+`;
+
+export const modalCss = css`
+  width: 100%;
+  max-width: 700px;
+  margin: 70px auto;
+`;
+
+export default withApi(withGlobalSelection(withTags(DashboardWidgetQuerySelectorModal)));
```
[the remark this answers is not in the document]
Comment (edwardgou-sentry) [reply static/app/components/modals/dashboardWidgetQuerySelectorModal.tsx]: Removed!

File: static/app/components/modals/dashboardWidgetQuerySelectorModal.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    const {organization, widget, selection} = this.props;
+    const querySearchBars = widget.queries.map(query => {
+      const eventView = eventViewFromWidget(
+        widget.title,
+        query,
+        selection,
+        widget.displayType
+      );
+      return (
+        <React.Fragment key={query.id}>
```
[the remark this answers is not in the document]
Comment (edwardgou-sentry) [reply static/app/components/modals/dashboardWidgetQuerySelectorModal.tsx]: Good call, I've updated to use index instead!

File: static/app/views/dashboardsV2/widgetCard.tsx (hunk 1 above)
Comment (edwardgou-sentry) [reply static/app/views/dashboardsV2/widgetCard.tsx]: I think you're right, I've updated to just redirect the user to Discover when the widget only has 1 query. Since table widgets only have 1 query, the redirect should be immediate without the modal pop up.

File: tests/js/spec/components/modals/dashboardWidgetQuerySelectorModal.spec.tsx (hunk 2 above)
Comment (edwardgou-sentry) [reply tests/js/spec/components/modals/dashboardWidgetQuerySelectorModal.spec.tsx]: fixed!

File: static/app/views/dashboardsV2/types.tsx [hunk 6]
```diff
@@ -11,6 +11,7 @@ export enum DisplayType {
 }
 
 export type WidgetQuery = {
+  id?: string;
```
Comment (shruthilayaj) [inline static/app/views/dashboardsV2/types.tsx:14]: Is this change still required?

**Review by shruthilayaj — APPROVED** [review]
one small comment, but otherwise LGTM!

File: static/app/views/dashboardsV2/types.tsx (hunk 6 above)
Comment (edwardgou-sentry) [reply static/app/views/dashboardsV2/types.tsx]: Nope! No longer required after we switched to index instead of id for key. Updated!
